runtime/win32.c

```diff
@@ -431,7 +431,7 @@ CAMLexport int caml_read_directory(wchar_t * dirname,
   h = _wfindfirst(template, &fileinfo);
   if (h == -1) {
     caml_stat_free(template);
-    return errno == ENOENT ? 0 : -1;
+    return -1;
```

Comment:
Unfortunately root directories can be empty, even though it's not that likely. There are two possible fixes: the dirty fix is to rely on the fact that `GetLastError` is not cleared by `_wfindfirst`:
```suggestion
    return GetLastError() == ERROR_FILE_NOT_FOUND ? 0 : -1;
```
but this isn't satisfying - especially as in this case you get `ERROR_FILE_NOT_FOUND` as opposed to the more expected `ERROR_NO_MORE_FILES`. However, the alternative is simply to "stat" it at this point. The classic (see, of course, [Raymond Chen](https://devblogs.microsoft.com/oldnewthing/20071023-00/?p=24713)) way to do that is to use `GetFileAttributes`:
```suggestion
    return errno == ENOENT && GetFileAttributes(dirname) != INVALID_FILE_ATTRIBUTES ? 0 : -1;
```
I tested with a manually cleared USB stick locally, and both of those correctly return `[||]` for my `G:\` but now correctly raise an exception for the un-mapped `X:\`